### PR TITLE
gh-89039: call subclass constructors in datetime.*.replace

### PR DESCRIPTION
**Pull request by eltoder**
When replace() method is called on a subclass of datetime, date or time, properly call derived constructor. Previously, only the base class's constructor was called.

Also, make sure to pass non-zero fold values when creating subclasses in various methods. Previously, fold was silently ignored.

* Issue: gh-89039

### CODE REVIEW

**Comment by eltoder** [comment]
@serhiy-storchaka This is a follow-up to #112921 to fix another issue. Would appreciate if you could take a look.

File: Modules/_datetimemodule.c
Comment (serhiy-storchaka) [inline Modules/_datetimemodule.c]: It is confusing that `res` is used for two different variables. I suggest to rename it to `rc` (return code). Or get rid of it and duplicate `Py_DECREF` in the `if` branch, what you prefer.

File: Modules/_datetimemodule.c
Comment (serhiy-storchaka) [inline Modules/_datetimemodule.c]: PEP 7 nitpick:
```suggestion
    }
    else {
```

File: Modules/_datetimemodule.c
Comment (serhiy-storchaka) [inline Modules/_datetimemodule.c]: PEP 7.
```suggestion
call_subclass_fold(PyObject *cls, int fold, const char *format, ...)
{
```

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM. I have only few style nitpicks.

File: Modules/_datetimemodule.c
Comment (eltoder) [reply Modules/_datetimemodule.c]: I did a quick grep, and `err` seems like a more popular choice than `rc`.

File: Modules/_datetimemodule.c
Comment (eltoder) [reply Modules/_datetimemodule.c]: Done.

File: Modules/_datetimemodule.c
Comment (eltoder) [reply Modules/_datetimemodule.c]: Done.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Thank you for your contribution @eltoder. Do you mind to create a backport for 3.12?

**Comment by eltoder** [comment]
@serhiy-storchaka Thank you for merging. Can we just backport #112921? Then this will work automatically. Or should I write a version of this change that doesn't use argument clinic?

**Comment by eltoder** [comment]
@serhiy-storchaka The conflicts were very easy to resolve. I did it here: https://github.com/python/cpython/pull/115344

But I get a complaint from CI that the ABI changed. Is that correct or a false positive?

**Comment by serhiy-storchaka** [comment]
It does not allow to merge the PR without sanction of the Release Manager.

**Comment by eltoder** [comment]
@serhiy-storchaka What would you recommend? The backport itself is trivial, but it does change the size of pyruntimestate. Does this really affect the ABI? I see that there were some updates to it on the 3.12 branch by @ericsnowcurrently after 3.12 was released: https://github.com/python/cpython/commits/3.12/Doc/data/python3.12.abi
Maybe pyruntimestate is not really a part of the ABI?

**Comment by eltoder** [comment]
@ericsnowcurrently Do you think it is acceptable to change pyruntimestate to backport #112921? I used argument clinic and it added some new string constants.

**Comment by ericsnowcurrently** [comment]
I don't see a problem with it, but release manager makes the final decision.

**Pull request merged**
